contracts/AelinPool.sol

```diff
@@ -385,6 +443,103 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
     }
 
+    /**
+     * @dev allows anyone to become a purchaser with a qualified erc721
+     * nft in the pool depending on the scenarios
+     *
+     * Scenarios:
+     * 1. each wallet holding a qualified NFT to deposit an unlimited amount of purchase tokens
+     * 2. certain amount of purchase tokens per wallet regardless of the number of qualified NFTs held
+     * 3. certain amount of Investment tokens per qualified NFT held
+     */
+
+    function purchasePoolTokensWithNft(NftArgs[] calldata nftArgs, uint256 _purchaseTokenAmount) external lock {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+
+        uint256 maxPurchaseTokenAmount;
+
+        for (uint256 i = 0; i < nftArgs.length; i++) {
+            NftArgs memory tmpNftArg = nftArgs[i];
+            address _collectionAddress = tmpNftArg.collectionAddress;
+            uint256[] memory _tokenIds = tmpNftArg.tokenIds;
+
+            NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+            require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+
+            if (tmpNftData.purchaseAmountPerToken) {
+                maxPurchaseTokenAmount += tmpNftData.purchaseAmount * _tokenIds.length;
+            }
+
+            if (!tmpNftData.purchaseAmountPerToken && tmpNftData.purchaseAmount > 0) {
+                require(!nftWalletUsedForPurchase[_collectionAddress][msg.sender], "wallet already used for collection");
+                nftWalletUsedForPurchase[_collectionAddress][msg.sender] = true;
+                maxPurchaseTokenAmount += tmpNftData.purchaseAmount;
+            }
+
+            if (tmpNftData.purchaseAmount == 0) {
+                maxPurchaseTokenAmount = _purchaseTokenAmount;
+            }
+
+            if (NftCheck.supports721(_collectionAddress)) {
+                _purchasePoolWith721(_collectionAddress, _tokenIds);
+            }
+            if (NftCheck.supports1155(_collectionAddress)) {
+                _purchasePoolWith1155(_collectionAddress, _tokenIds, tmpNftData);
+            }
+            if (_collectionAddress == CRYPTO_PUNKS) {
+                _purchasePoolWithPunks(_collectionAddress, _tokenIds);
+            }
+        }
+
+        require(_purchaseTokenAmount <= maxPurchaseTokenAmount, "purchase amount should be less the max allocation");
+
+        uint256 amountBefore = IERC20(purchaseToken).balanceOf(address(this));
+        IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        uint256 amountAfter = IERC20(purchaseToken).balanceOf(address(this));
+        uint256 purchaseTokenAmount = amountAfter - amountBefore;
+
+        if (purchaseTokenCap > 0) {
+            uint256 totalPoolAfter = totalSupply() + purchaseTokenAmount;
+            require(totalPoolAfter <= purchaseTokenCap, "cap has been exceeded");
```

Comment:
this is correct. I wonder if some people will have high gas costs from all the loops only to find out that they exceeded the pool cap at the end of the method. it should not happen often in any case.